Connected generate schedules

gannonprudhomme opened the pull request:
Connects the `Generate schedules` button to /scheduler/generate.

Currently I handle the BASIC/SECTIONS customization level by if the SECTIONS option is selected, then the options for the honors/web only are set to the default. And if the BASIC option is selected, then the sections aren't sent. We might want to modify this for sessions so we *always* have the data in order to save it, and we can instead just send the current customization level for each section, and `create_schedules` can then choose which one it wants to ignore. Should be a relatively easy fix.

Also think we might want to add some tests for this that I missed, as the availabilities time converting was messed up and I didn't catch it for awhile, so lmk if you have any ideas for tests to add.

I also had to change the course cards actions which sort of (not really) reverts the changes Ryan made in #244 since I actually do need the original `parseMeetings` functionality for this, but I did my best extracting-out the common functionality of it all.

Closes #193

firejake308 commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -36,57 +36,98 @@ function updateCourseCardSync(index: number, courseCard: CourseCardOptions): Upd
+}
+
+/**
+ * Helper function that parses all of the meetings in the section data and returns them.
```
Based on the docstrings, `parseAllMeetings` and `parseMeetings` seem to do the same thing. Could you clarify that a bit?

firejake308 commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -36,57 +36,98 @@ function updateCourseCardSync(index: number, courseCard: CourseCardOptions): Upd
  *  @param arr The array of sections returned from the backend, such as from api/sections
  */
-export function parseSections(arr: any[]): SectionSelected[] {
+export function parseSections(arr: any[]): SectionSelected[] { // Rename to parseSectionSelected?
```
I think we probably should rename it, just so that no one is surprised by the return type

firejake308 commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -36,57 +36,98 @@ function updateCourseCardSync(index: number, courseCard: CourseCardOptions): Upd
-    sectionData.meetings.forEach((meetingData: any) => {
-      let start: string[] = ['0', '0'];
-      let end: string[] = ['0', '0'];
+    const section = parseSection(sectionData);
```
So basically, these two functions have been separated out into helper functions, and those helper functions are re-used between `parseSections` and `parseAllMeetings`, right?

gannonprudhomme replied:
Yeah exactly. They're only very slightly different and since they're both needed for different situations, I thought extracting out the common functionality was the best move

firejake308 commented on autoscheduler/frontend/src/tests/ui/ConfigureCard.test.tsx:
```diff
@@ -1,19 +1,42 @@
+import { updateCourseCard } from '../../redux/actions/courseCards';
+import { CustomizationLevel } from '../../types/CourseCardOptions';
 
 jest.mock('../../components/SchedulingPage/ConfigureCard/generateSchedulesMock', () => ({
```
We can probably remove this mock

firejake308 commented on autoscheduler/frontend/src/tests/ui/ConfigureCard.test.tsx:
```diff
@@ -1,19 +1,42 @@
     test('when the user clicks Fetch Schedules', () => {
+      let fetchCount = 0;
+
+      fetchMock.mockImplementation((route: string): Promise<Response> => {
```
If there are no other calls to `fetch` in this test, we might be able to `expect(fetchMock).toBeCalledWith('/scheduler/generate')`

gannonprudhomme replied:
So I thought this would work too, but since it's a POST request with a JSON body, we'd also have to assert the body of it:

```
 FAIL  tests/ui/ConfigureCard.test.tsx (6.81s)
  ● ConfigureCard component › makes an API call › when the user clicks Fetch Schedules

    expect(jest.fn()).toHaveBeenCalledWith(...expected)

    Expected: "scheduler/generate"
    Received: "scheduler/generate", {"body": "{\"term\":\"\",\"includeFull\":false,\"courses\":[],\"availabilities\":[]}", "method": "POST"}

    Number of calls: 1
```

firejake308 replied:
In that case, try `expect(fetchMock).toBeCalledWith('/scheduler/generate', expect.any())`

firejake308 commented on autoscheduler/scheduler/views.py:
```diff
@@ -18,10 +18,10 @@ def _parse_course_filter(course) -> CourseFilter:
 
-    honors = BasicFilter(course["honors"])
-    web = BasicFilter(course["web"])
+    honors = BasicFilter(course.get("honors"))
```
Is `.get()` more reliable, or are you just using it to provide a default value?

gannonprudhomme replied:
So the `[key]` accessing will throw an error if the key doesn't exist in the data, whereas `.get()` returns `None` if it doesn't exist. This way if `honors`/`web` isn't in the sent data it will still return stuff. This case shouldn't ever actually occur, but it's probably a better practice to use `.get()`

firejake308 commented on autoscheduler/frontend/src/tests/ui/ConfigureCard.test.tsx:
```diff
@@ -26,7 +49,74 @@ describe('OptionsCard component', () => {
+      const { sections } = JSON.parse(body.toString());
+
+      // assert
+      expect(sections).toBeUndefined();
```
Based on my reading of `ConfigureCard.tsx`, shouldn't `sections` be an empty array rather than undefined?

gannonprudhomme replied:
Yeah that's a great point. I think I ran into trouble when I tried to make it an empty array that it would be `undefined` anyways, but I'll double check and get back to you

gannonprudhomme commented:
- Rebased with `master` and resolved conflicts
- Squashed the `parseAllMeetings` and `parseMeeting` docstring updates and the `sectionDAta` type into `Refactored parseSections and added parseAllMeetings`
- Renamed `parseSection` to `parseSectionSelected`
- Added loading indicator mocking & removed unused mock

I'll look into the `undefined` section assertions then will re-request review afterwards

gannonprudhomme commented:
Okay so turns out I had my test set up like completely wrong. Will try to get them fixed before the meeting tomorrow, but I think I'm going to go insane if I work on them for any longer right now

ryan-conn commented on autoscheduler/frontend/src/timeUtil.ts:
```diff
@@ -8,7 +8,13 @@ export const LAST_HOUR = 21;
-export function formatTime(h: number, m: number, use24Hour = false): string {
-  return `${use24Hour ? h : (h - 1) % 12 + 1}:${
+export function formatTime(h: number, m: number, use24Hour = false, padZeroes = false): string {
+  const ret = `${use24Hour ? h : (h - 1) % 12 + 1}:${
     new Intl.NumberFormat('en-US', { minimumIntegerDigits: 2 }).format(m)}`;
+
+  if (padZeroes && ret.length < 4) {
+    return `0${ret}`;
+  }
+
+  return ret;
```
This function is a bit messy right now, if you reuse the NumberFormat for hours it's more straightforward:
```suggestion
  const formattedHours = use24Hour ? h : (h - 1) % 12 + 1;
  const padZero = new Intl.NumberFormat('en-US', { minimumIntegerDigits: 2 });
  return `${padZeroes ? padZero.format(formattedHours) : formattedHours}:${padZero.format(m)}`;
```

ryan-conn commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -36,57 +36,99 @@ function updateCourseCardSync(index: number, courseCard: CourseCardOptions): Upd
+ * @param section The parsed Section type from this sectionData
+ */
+function parseMeetings(sectionData: any, section: Section): Meeting[] {
+  const meetings: Meeting[] = [];
+
+  sectionData.meetings.forEach((meetingData: any) => {
+    let start: string[] = ['0', '0'];
+    let end: string[] = ['0', '0'];
+
+    if (meetingData.start_time != null && meetingData.start_time.length > 0) {
+      start = meetingData.start_time.split(':');
+      end = meetingData.end_time.split(':');
+    }
+
+    const meeting = new Meeting({
+      id: Number(meetingData.id),
+      building: '',
+      meetingDays: meetingData.days,
+      startTimeHours: Number(start[0]),
+      startTimeMinutes: Number(start[1]),
+      endTimeHours: Number(end[0]),
+      endTimeMinutes: Number(end[1]),
+      // Converts string to MeetingType enum
+      meetingType: MeetingType[meetingData.type as keyof typeof MeetingType],
+      section,
+    });
+
+    meetings.push(meeting);
+  });
+
+  return meetings;
```
Since all the `forEach` is doing is appending a single item to an array, you can use `sectionData.meetings.map()` instead

ryan-conn commented on autoscheduler/frontend/src/redux/actions/courseCards.ts:
```diff
@@ -36,57 +36,99 @@ function updateCourseCardSync(index: number, courseCard: CourseCardOptions): Upd
+export function parseSectionSelected(arr: any[]): SectionSelected[] {
   const sections: SectionSelected[] = [];
 
   arr.forEach((sectionData) => {
```
Change this to use `arr.map()`

firejake308 commented on autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx:
```diff
@@ -35,39 +36,58 @@ const ConfigureCard: React.FC = () => {
+        const selectedSections: string[] = []; // the section nums of the selected sections
+
+        // Iterate through the sections and only choose the ones that are selected
+        courseCard.sections.forEach((sectionSel) => {
```
This could probably be implemented as a `.filter()`

firejake308 commented on autoscheduler/frontend/src/components/SchedulingPage/ConfigureCard/ConfigureCard.tsx:
```diff
@@ -35,39 +36,58 @@ const ConfigureCard: React.FC = () => {
-          web: courseCards[i].web,
+          sections: isBasic ? [] : selectedSections, // Only send if "Section" customization level
+          // Only send if "Basic" level
+          honors: isBasic ? courseCard.honors ?? filterDefault : filterDefault,
```
Can I get some parentheses around this to make it easier to read? Side note, nice use of the new `??` operator

gannonprudhomme commented:
> The first time I ran it, I forgot to check "include full sections" since most sections are full now, so I'd like to have an error message when there are no schedules. We could consider a [snackbar](https://material-ui.com/components/snackbars/), a dialog, or just a message in the `SchedulePreview`, but something owuld be nice.

Yeah good call, I think a snackbar would work well. Would rather do this in a separate PR though honestly, as I just want this merged.

> Also, this would be a backend change I guess, but maybe the API route should have a query/body parameter for the number of schedules to return in case we want more than 5 later?

It's already an option for the actual `create_schedules` function so this would be really easy. Would like to do this in a separate PR though whenever we actually want to, as again it'd be a really small change. We can discuss during the meeting though
